### PR TITLE
Traffic

### LINKED ISSUE #42: Use Pygal
> ### Redmine metadata
> 
> | Attribute | Value |
> | --- | --- |
> | done_ratio | 0 |
> | status | New |
> | start_date | 2015/10/09 |
> | created_on | 2015/10/09 02:33:38 +0200 |
> | author | @lukasjuhrich |
> | priority | Normal |
> | updated_on | 2016/09/18 19:50:36 +0200 |
> | tracker | Feature |
> | asigned_to | None |

---

One could use PyGal for the traffic charts instead of the current Highcharts solution, as it works really fine in sipa and is easy to implement.
## 

_This issue has been automatically imported from redmine on 2016-09-28 19:09:59.127060_

### PR DESCRIPTION
**Pull request by mueslo**
Change traffic model to a model based on three parts:
  * `TrafficBalance` which *uniquely* contains the balance of the user at some timestamp
  * `TrafficVolume` which contains timestamped traffic usage events
  * `TrafficCredit` which contains timestamped traffic credit events

Replace highcharts traffic chart with a custom d3.js one (obsoletes ergo closes #42)
trafficgraph

Open questions/TODOs
- [ ] Possibly also allow/properly handle future timestamps
- [ ] Remove Group inheritance scheme, so a Group can have both Properties and Traffic associated with it.

### CODE REVIEW

**Comment by lukasjuhrich** [comment]
Thanks for the quick response, I will have a look at it tomoroow when I have time.

Looks very good! How did you do the animation?

**Comment by lukasjuhrich** [comment]
@mueslo What are the things you intend to finish? Are there any blockers? If appropriate, consider adding a TODO-List in the PR description.

Also, perhaps you'd like to rebase, as a lot of new commits have been added recently.

**Comment by mueslo** [comment]
Should be ready to be merged now (everything should be cleaned up and working at least as well as before, any further changes can be done on the main branch), although I'm not sure whether @sebschrader still has any issues with it.

File: pycroft/model/user.py [hunk 1]
```diff
@@ -78,7 +78,8 @@ class User(ModelBase, UserMixin):
         # but between punctuation characters must be characters or numbers
         (?:[._-]?[a-z0-9])+$
         """, re.VERBOSE)
-    email_regex = re.compile(r"^[a-zA-Z0-9]+(?:(?:\+|-|_|\.)[a-zA-Z0-9]+)*"
+    email_regex = re.compile(r"^[a-zA-Z0-9!#$%&'*+\-/=?^_`{|}~]+"
```
Comment (lukasjuhrich) [inline pycroft/model/user.py]: I'm not sure whether we should allow such email addresses. Wasn't it the case that this regex has been designed rather simple on purpose? For instance, email addresses with two subsequent underscores are invalid as well.
I  know the address you're referring to, and @lightning1 told me it was probably a mistake.

Comment (mueslo) [inline pycroft/model/user.py]: Are invalid according to what?

**Comment by lukasjuhrich** [comment]
@mueslo I remember something that's missing: Importing the traffic data!

Do you/we want to fix this in this branch, or would you like a separate issue/PR for this? I would stand for the latter as merging this branch would prevent a further delay of getting the model changes to the main branch (develop).

**Comment by lukasjuhrich** [comment]
I will merge this PR as everything seems fine by me, especially in the model (although I haven't tested the animations to be honest).

**Pull request merged**